src/librustc_codegen_ssa/back/write.rs

```diff
@@ -1022,7 +1022,13 @@ fn start_executing_work<B: ExtraBackendMethods>(
         None
     };
 
-    let ol = tcx.backend_optimization_level(LOCAL_CRATE);
+    let ol = if tcx.sess.opts.debugging_opts.no_codegen
+             || !tcx.sess.opts.output_types.should_codegen() {
+        // If we know that we won’t be doing codegen, create target machines without optimisation.
+        config::OptLevel::No
+    } else {
+        tcx.backend_optimization_level(LOCAL_CRATE)
+    };
```

Comment:
Is there a reason we're not putting this "don't optimize if no codegen" into the backend_optimization_level query itself?

Reply:
Ideally a fix would have happened all the way down inside `collect_and_partition_mono_items` (which would take care to not ICE when no codegen is happening), however that change is more involved than this and, given that I haven’t setup myself for reproduction, I wasn’t in a place to make such a change.